artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/AddressImpl.java

```diff
@@ -84,6 +84,7 @@ public void addLinkedAddress(final Address address) {
       if (linkedAddresses == null) {
          linkedAddresses = PlatformDependent.hasUnsafe() ? new NonBlockingHashSet<>() : new ConcurrentHashSet<>();
       }
+      assert !linkedAddresses.contains(address);
```

Comment:
Previously if it called twice it wont error but simply proceed. Is this really a good idea, risk breaking or erroring

Reply:
Im not feeling comfortable about this.

Reply:
Better to change return as boolean and simply return true or false if succeeded and if you care if it succeeded use that to subsequently handle.

Reply:
I like this solution!

Reply:
I am not a fan of asserts generally, but in this case, if I added that assert with my original change, this PR would not be necessary. I think it serves a purpose. It forces a test failure on bad usage.

Reply:
see the original pr: https://github.com/apache/activemq-artemis/pull/3344#discussion_r524344331

Reply:
But that is at risk at causing a runtime issue in real systems. If you want to validate it, as noted and as per general collection kind of behaviour its normal to return true false if a change occured not error

Reply:
I'm not advocating one way or the other here, but aren't assertions disabled by the JVM by default and must be enabled using `-ea`?

Reply:
In this case what could be the proper action in case we see the linked address to be added twice? A log? An exception?
I suppose both are poor to capture a performance need (more then a correctness thing). I am open to proposals here but I am not sure asserts are not good if is to guide specific invariant to be checked with ki cost at runtime, especially for performance reasons.

Reply:
Said that I think is not bad to return a Boolean but we still need to under what action to take in case of false branch: If not, I don't see a reason to add it to the API. Differently from a collection, that's a generic reusable piece of code, this one is domain code and isn't there to be reusable.

Reply:
@jbertram yep and iirc surefire add it by default while running tests, that's why I have suggested that option

Reply:
We shouldnt be relying on code that only does something in tests as then at runtime we have different behaviour.

Reply:
If you need something just for tests we would be better extracting interface and then in test injecting a mock or interception that checks the call count.

Reply:
in this case, the different behaviour is correct, just a waste. The code now does the add/remove of linked addresses once, and if someone changes that, the assert will fire. When I made that optimisation without the assert, it was dropped in a few commits and none the wiser. an assert is a good win here imho. It is suggesting the contract but it is not a fatal flaw to insert the same value twice, just a waste.
@michaelandrepearce are you -1 on asserts or just this assert ?

Reply:
to progress this, the assert is gone!